quickwit/quickwit-indexing/src/source/queue_sources/visibility.rs

```diff
@@ -0,0 +1,226 @@
+}
+
+/// Spawns actor that ensures that the visibility of a given message
+/// (represented by its ack_id) is extended when required. We prefer applying
+/// ample margins in the extension process to avoid missing deadlines while also
+/// keeping the number of extension requests (and associated cost) small.
+pub(super) fn spawn_visibility_task(
+    ctx: &SourceContext,
+    queue: Arc<dyn Queue>,
+    ack_id: String,
+    current_deadline: Instant,
+    visibility_settings: VisibilitySettings,
+) -> VisibilityTaskHandle {
+    let task = VisibilityTask {
+        queue,
+        ack_id: ack_id.clone(),
+        extension_count: 0,
+        current_deadline,
+        stop_extension_loop: false,
+        visibility_settings,
+    };
+    let (_mailbox, _actor_handle) = ctx.spawn_actor().spawn(task);
+    VisibilityTaskHandle {
+        mailbox: _mailbox,
+        actor_handle: _actor_handle,
+        ack_id,
+    }
+}
+
+impl VisibilityTask {
+    async fn extend_visibility(
+        &mut self,
+        ctx: &ActorContext<Self>,
+        extension: Duration,
+    ) -> anyhow::Result<()> {
+        let _zone = ctx.protect_zone();
+        self.current_deadline = tokio::time::timeout(
+            self.visibility_settings.request_timeout,
+            self.queue.modify_deadlines(&self.ack_id, extension),
+        )
+        .await
+        .context("deadline extension timed out")??;
+        self.extension_count += 1;
+        Ok(())
+    }
+
+    fn next_extension(&self) -> Duration {
+        (self.current_deadline - Instant::now())
+            - self.visibility_settings.request_timeout
+            - self.visibility_settings.request_margin
+    }
+}
+
+impl VisibilityTaskHandle {
+    pub fn extension_failed(&self) -> bool {
+        self.actor_handle.state() == ActorState::Failure
+    }
+
+    pub fn ack_id(&self) -> &str {
+        &self.ack_id
+    }
+
+    pub async fn last_extension(self, extension: Duration) -> anyhow::Result<()> {
+        self.mailbox
+            .ask_for_res(RequestLastExtension { extension })
+            .await
+            .map_err(|e| anyhow!(e))?;
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl Actor for VisibilityTask {
+    type ObservableState = JsonValue;
+
+    fn name(&self) -> String {
+        "QueueVisibilityTask".to_string()
+    }
+
+    async fn initialize(&mut self, ctx: &ActorContext<Self>) -> Result<(), ActorExitStatus> {
+        let first_extension = self.next_extension();
+        if first_extension.is_zero() {
+            return Err(anyhow!("initial visibility deadline insufficient").into());
+        }
+        ctx.schedule_self_msg(first_extension, Loop);
+        Ok(())
+    }
+
+    fn yield_after_each_message(&self) -> bool {
+        false
+    }
+
+    fn observable_state(&self) -> Self::ObservableState {
+        json!({
+            "ack_id": self.ack_id,
+            "extension_count": self.extension_count,
+        })
+    }
+}
+
+#[derive(Debug)]
+pub(super) struct Loop;
```

Comment:
```suggestion
struct Loop;
```

It is super important to make it private for correctness reasons.

(`Loop` messages are extremely dangerous, it is very important to make their proofreading as trivial as possible)